crates/runix/src/installable.rs

```diff
@@ -1,45 +1,285 @@
+        if !(VALID_ATTRIBUTE.is_match(s)) {
+            Err(ParseInstallableError::InvalidAttr(s.to_string()))?;
+        }
+        Ok(Attribute(s.to_string()))
+    }
+}
+
+impl Display for Attribute {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.0.contains('.') {
+            write!(f, "\"{}\"", self.0)
+        } else {
+            write!(f, "{}", self.0)
+        }
+    }
+}
+
+impl FromStr for Installable {
+    type Err = ParseInstallableError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.split_once('#') {
+            Some((flakeref, attr_path)) => Ok(Installable {
+                flakeref: flakeref.parse()?,
+                attr_path: attr_path.parse()?,
+            }),
+            None => Ok(Installable {
+                flakeref: s.parse()?,
+                attr_path: "".parse()?,
+            }),
+        }
+    }
+}
+
+impl Display for Installable {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.flakeref)?;
+        if !self.attr_path.is_empty() {
+            write!(f, "#{}", self.attr_path)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ParseInstallableError {
+    #[error(transparent)]
+    ParseFlakeRef(#[from] ParseFlakeRefError),
+    #[error("Installable is missing an attribute path")]
+    MissingAttrPath,
+    #[error("Invalid attrpath encoding: {0}")]
+    FragmentEncoding(#[from] std::str::Utf8Error),
+    #[error("Missing closing quote in selection path: '{0}'")]
+    UnclosedQuote(String),
+    #[error("Invalid attribute '{0}'")]
+    InvalidAttr(String),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_parse_as(input: &str, expected: &str, description: &str) {
+        let output = input.parse::<AttrPath>().expect(description);
+        assert_eq!(output.to_string(), expected);
+    }
+
+    fn assert_parse(input: &str, description: &str) {
+        let output = input.parse::<AttrPath>().expect(description);
+        assert_eq!(output.to_string(), input);
+    }
+
+    fn assert_parse_compontents<'a>(
+        components: impl AsRef<[&'a str]>,
+        expect: &str,
+        description: &str,
+    ) {
+        let output = AttrPath::try_from(components.as_ref()).expect(description);
+        assert_eq!(output.to_string(), expect);
+    }
+
+    fn assert_fail(input: &str, description: &str) {
+        input
+            .parse::<AttrPath>()
+            .expect_err(&format!("({input}) was expected to fail: {description}"));
+    }
+
+    #[test]
+    fn attr_path_from_str() {
+        assert_parse("a", "parse single attribute");
+        assert_parse("a.b.c", "parse nested path");
+        assert_parse_as("\"a\"", "a", "parse quoted single");
+        assert_parse("''a''", "parse '' quoted single");
+        assert_parse("2", "parse number");
+        assert_parse("\"a.b\"", "parse quoted dot");
```

Comment:
How come more of these aren't `assert_parse_as`?

Reply:
Those inputs produce the same output.
I wanted to highlight somehow the places where we do expect a change in the output, i.e. it will undo quoting if it's not required.
The other asserts are like the roundtrip tests in flakeref, they just check if the out put and the input are equal